src/components/application_manager/src/resumption/resumption_data_processor.cc

```diff
@@ -0,0 +1,1125 @@
+
+  for (auto& extension : application->Extensions()) {
+    extension->ProcessResumption(
+        saved_app,
+        [this](const int32_t app_id, const ResumptionRequest request) {
+          this->SubscribeToResponse(app_id, request);
+        });
+  }
+}
+
+void ResumptionDataProcessor::DeleteSubscriptions(
+    ApplicationSharedPtr application) {
+  LOG4CXX_AUTO_TRACE(logger_);
+  DeleteButtonsSubscriptions(application);
+  DeletePluginsSubscriptions(application);
+}
+
+void ResumptionDataProcessor::DeleteButtonsSubscriptions(
+    ApplicationSharedPtr application) {
+  LOG4CXX_AUTO_TRACE(logger_);
+  const ButtonSubscriptions button_subscriptions =
+      application->SubscribedButtons().GetData();
+  for (auto& btn : button_subscriptions) {
+    const auto hmi_btn = static_cast<hmi_apis::Common_ButtonName::eType>(btn);
+    if (hmi_apis::Common_ButtonName::CUSTOM_BUTTON == hmi_btn) {
+      continue;
+    }
+    auto notification = MessageHelper::CreateOnButtonSubscriptionNotification(
+        application->hmi_app_id(), hmi_btn, false);
+    // is_subscribed = false
+    ProcessMessageToHMI(notification, false);
+    application->UnsubscribeFromButton(btn);
+  }
+}
+
+void ResumptionDataProcessor::DeleteWindowsSubscriptions(
+    ApplicationSharedPtr application) {
+  LOG4CXX_AUTO_TRACE(logger_);
+
+  const auto window_ids = application->GetWindowIds();
+  for (const auto& window_id : window_ids) {
+    const auto hmi_state = application->CurrentHmiState(window_id);
+    if (mobile_apis::WindowType::WIDGET != hmi_state->window_type()) {
+      continue;
+    }
+
+    LOG4CXX_DEBUG(logger_, "Reverting CreateWindow for: " << window_id);
+
+    auto delete_request = MessageHelper::CreateUIDeleteWindowRequestToHMI(
+        application, application_manager_, window_id);
+    const bool subscribe_on_request_events = false;
+    ProcessMessageToHMI(delete_request, subscribe_on_request_events);
+
+    application->RemoveWindowInfo(window_id);
+    application->RemoveHMIState(window_id,
+                                app_mngr::HmiState::StateID::STATE_ID_REGULAR);
+    application->remove_window_capability(window_id);
+  }
+}
+
+void ResumptionDataProcessor::DeletePluginsSubscriptions(
+    application_manager::ApplicationSharedPtr application) {
+  LOG4CXX_AUTO_TRACE(logger_);
+
+  resumption_status_lock_.AcquireForReading();
+  auto it = resumption_status_.find(application->app_id());
+  if (it == resumption_status_.end()) {
+    resumption_status_lock_.Release();
+    return;
+  }
+
+  const ApplicationResumptionStatus& status = it->second;
+  smart_objects::SmartObject extension_subscriptions;
+  for (auto ivi : status.successful_vehicle_data_subscriptions_) {
+    LOG4CXX_DEBUG(logger_, "ivi " << ivi << " should be deleted");
+    extension_subscriptions[ivi] = true;
+  }
+  resumption_status_lock_.Release();
+
+  for (auto& extension : application->Extensions()) {
+    extension->RevertResumption(extension_subscriptions);
```

Comment:
@iCollin you are right, but please note that `extension_subscriptions` is the smart object which is a dynamic object. That means it may contain any structure of any information required by all other extensions. In terms of that feature, current information is enough for all covered extensions.
However in a future implementations the structure of that object can be updated to provide additional information for the rest of extensions. As a good example, please take a look at 0188 feature implementation - this function was updated and smart object contains more information to cover more extensions.

Reply:
okay thank you for explaining. we will also need to add data for reverting app service data subscriptions in this function as well, maybe it will be easiest to do that in a separate PR based on 0188.